Trac3215

rpetry opened the pull request:
Re-implemented find.best.design.elements() in Java
@olgamelnichuk, @alf239 - could you review please

rpetry committed:
Ticket #3215 - Re-implemented find.best.design.elements() in Java

rpetry committed:
Merge branch 'develop' into trac3215

rpetry committed:
Ticket #3215 - Removed no longer used classes

rpetry committed:
Ticket #3215 - Re-generated serialVersionUID

rpetry committed:
Alas had to replicate compareTo() logic in BestDesignElementCandidate and PTRank due to memory problems during bitstats index build

rpetry committed:
Ticket #3215 - Alas had to replicate compareTo() logic in BestDesignElementCandidate and PTRank due to memory problems during bitstats index build

olgamelnichuk commented on atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentViewControllerBase.java:
```diff
@@ -47,6 +50,14 @@ class ExperimentViewControllerBase extends AtlasViewController {
+    public Map<Long, AtlasGene> getGenesByIds(List<Long> geneIds) {
+        return geneSolrDAO.getGenesByIds(geneIds);
+    }
+
```
What is the purpose of putting these methods and `geneSolrDAO` into base class? As I can see `ExperimentDesignViewController` doesn't need them..

rpetry replied:
You're right - moved them to `ExperimentViewController`.

olgamelnichuk commented on atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentViewController.java:
```diff
@@ -297,33 +292,44 @@ public class ExperimentViewController extends ExperimentViewControllerBase {
+            final List<Long> geneIds = !isNullOrEmpty(gid)
+                    ? findGeneIds(Arrays.asList(gid.trim()))
+                    : Collections.<Long>emptyList();
+
```
Is check for `!isNullOrEmpty(id)` really needed here?

rpetry replied:
would not `gid.trim()` throw an NPE if gid was null?

olgamelnichuk commented on atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentViewController.java:
```diff
@@ -297,33 +292,44 @@ public class ExperimentViewController extends ExperimentViewControllerBase {
+                        !isNullOrEmpty(ef) && !isNullOrEmpty(efv) ?
+                                // We don't currently allow search for best design elements by either just an ef
+                                // or just an efv - both need to be specified
+                               Collections.singleton(Pair.create(ef, efv)) : Collections.<Pair<String, String>>emptySet() ,
```
I prefer to see this comment in the doc of findBestGenesForExperiment..

rpetry replied:
I've added it there, as requested, but will leave it here also as we're performing that test just above.

olgamelnichuk commented on atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentViewController.java:
```diff
@@ -297,33 +292,44 @@ public class ExperimentViewController extends ExperimentViewControllerBase {
+                );
+            }
+
+            final Map<Long, AtlasGene> geneIdToGene = getGenesByIds(res.getGeneIds());
```
I'm still thinking that the service approach was more clear; it encapsulated all stuff it uses... but now you have to carry geneSolrDAO every time you need to use findBestGenesForExperiment.

rpetry replied:
findBestGenesForExperiment belongs in ExperimentWithData, along with all other methods searching the ncdfs.  The comment in AtlasExperimentAnalyticsViewService was:  'This class provides access to the statistical functions defined in R scripts' and yet it was coupling calls to R and Solr search. Now we bring ncdf search and Solr search at the point we need them together: ExperimentViewController and otherwise keep them apart. I'd prefer to keep things as they are now.

olgamelnichuk replied:
I used to think that the findBestGenesForExperiment is part of business logic, but not data layer.

olgamelnichuk commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/ExperimentWithData.java:
```diff
@@ -517,4 +521,138 @@ public class ExperimentWithData implements Closeable {
+            return EMPTY_DESIGN_ELEMENT_RESULT;
+
+        final BestDesignElementsResult result = new BestDesignElementsResult();
+        result.setArrayDesignAccession(expPart.getArrayDesign().getAccession());
```
A cyclic dependancy: ExperimentWithData -> ExperimentPart -> ExperimentWithData.  Looks like a bad design...

```
expPart.getArrayDesign().getAccession(), 
expPart.getGeneIds(), 
expPart.getUniqueEFVs(), 
expPart.getTStatistics(),
expPart.getPValues()
```

why not to put findBestGenes functionality into expPart?

rpetry replied:
You're right - it doesn't look good - moved it to expPart

olgamelnichuk commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/BestDesignElementCandidate.java:
```diff
@@ -0,0 +1,122 @@
+    Float tStat;
+    Integer deIndex;
+    Integer uEFVIndex;
+
```
Why using default access for class fields?

rpetry replied:
My bad - changed to private

olgamelnichuk commented on atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentViewController.java:
```diff
@@ -297,33 +292,44 @@ public class ExperimentViewController extends ExperimentViewControllerBase {
                     new Function<BestDesignElementsResult.Item, ExperimentTableRow>() {
                         public ExperimentTableRow apply(@Nonnull BestDesignElementsResult.Item item) {
-                            return new ExperimentTableRow(item);
+                            return new ExperimentTableRow(item, geneIdToGene.get(item.getGeneId()));
```
same problem here... item already should contain AtlasGene instance. not just geneId.

rpetry replied:
.. and the same answer - BestDesignElementsResult is returned by ncdf search functionality and has nothing to do with Solr - hence can't contain AtlasGene.

alf239 replied:
Add another layer. I cannot see why controller must collect data from all over the application. I can get it when there was an old and broken code, but here we take code working with proper objects, rip it apart, and switch back to ids. I do believe we can do better than that.

olgamelnichuk commented on atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentViewController.java:
```diff
@@ -423,9 +429,9 @@ public class ExperimentViewController extends ExperimentViewControllerBase {
+        public ExperimentTableRow(BestDesignElementsResult.Item item, AtlasGene gene) {
+            geneName = gene != null ? gene.getGeneName() : "";
+            geneIdentifier = gene != null ? gene.getGeneIdentifier() : "";
             deAccession = item.getDeAccession();
```
How `gene` could be `null` here? Need to change design to eliminate this case..

rpetry replied:
I can't recall a reason why I put it there, but on the other hand cannot image why I would have put it there without seeing an NPE. As you say, by design this should not happen - have changed to `LogUtil.createUnexpected` is `gene` is `null`

alf239 replied:
NPE is unexpected (and unchecked) enough, there's no need to clutter the code. If you can see an NPE, investigation is the way. Checks with no apparent reason all over the code carry reader's attention away and tend to grow. Please avoid.

rpetry replied:
I understand - have removed and will review if an NPE ever occurs

rpetry committed:
Ticket #3215 - Incoroprated code review comments

olgamelnichuk commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/BestDesignElementsResult.java:
```diff
@@ -81,12 +79,12 @@ public class BestDesignElementsResult implements Iterable<BestDesignElementsResu
 
     public static class Item {
-        private final AtlasGene gene;
+        private final Long geneId;
```
What was wrong with `AtlasGene gene` here? Why we are back to indexes and ids?

rpetry committed:
Re-instated AtlasExperimentAnalysticsViewService

rpetry committed:
Ticket #3215 - Re-instated AtlasExperimentAnalysticsViewService

rpetry committed:
Ticket #3215 - Removed unused function

rpetry committed:
Ticket #3215 - Removed out-of-scope formatting changes

rpetry committed:
Ticket #3215 - Removed superfluous ExperimentTableRow constructor argument

olgamelnichuk commented on atlas-web/src/main/java/ae3/dao/GeneSolrDAO.java:
```diff
@@ -23,6 +23,7 @@
 package ae3.dao;
 
 import ae3.model.AtlasGene;
+import com.google.common.base.Functions;
```
unused import?

rpetry replied:
Removed

olgamelnichuk commented on atlas-web/src/main/java/ae3/service/experiment/AtlasExperimentAnalyticsViewService.java:
```diff
@@ -1,188 +1,155 @@
+
+    private boolean statisticsQualify(
+            final UpDownCondition upDownCondition, float pValue, float tStatistic) {
+        if (upDownCondition != upDownCondition.CONDITION_ANY && !upDownCondition.apply(UpDownExpression.valueOf(pValue, tStatistic)))
```
Using static method access via instance reference?

rpetry replied:
fixed

olgamelnichuk commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/BestDesignElementCandidate.java:
```diff
@@ -0,0 +1,122 @@
+            return -1;
+
+        return compare(getPValue(), o.getPValue()); // lower pVals come first
+    }
```
It would be nice to have a unit test; at least for `.compareTo()` method..

rpetry replied:
Have added junit test - thanks

alf239 replied:
Extended it with `quickcheck`, fixed the most obvious problems.

rpetry replied:
whooa - thanks Alexey - that's one elaborate test - you've certainly shown us the way to do it properly in the future!

alf239 replied:
Well, if you need The Way, try http://www.scs.stanford.edu/11au-cs240h/notes/testing.html — it explains the reasoning behind QuickCheck and its clones. I've just needed a way to check the method, and had been looking for an opportunity to try the approach for a long time already.

olgamelnichuk commented on atlas-web/src/main/java/uk/ac/ebi/gxa/web/controller/ExperimentDesignViewController.java:
```diff
@@ -19,8 +20,8 @@
     @Autowired
-    public ExperimentDesignViewController(ExperimentSolrDAO solrDAO, AtlasDAO atlasDAO) {
-        super(solrDAO, atlasDAO);
+    public ExperimentDesignViewController(ExperimentSolrDAO experimentSolrDAO, GeneSolrDAO geneSolrDAO, AtlasDAO atlasDAO) {
```
`geneSolrDAO` is not used here?

rpetry replied:
you're right - removed

rpetry committed:
Ticket #3215 - Code review comments

alf239 commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/BestDesignElementCandidate.java:
```diff
@@ -0,0 +1,110 @@
+        if (compare(getPValue(), o.getPValue()) == 0)
+            // if pvals are the same, return the this PT first
+            // (arbitrary which we return - it's just that one of them has to come first)
+            return -1;
```
In other words, we can have a case when `a < b`, and at the same time `b < a`? 

Doesn't sound good to me, and breaks the contract between `equals` and `compareTo`.

alf239 replied:
It is true that 

>  It is strongly recommended, but _not_ strictly required that `(x.compareTo(y)==0) == (x.equals(y))`.  Generally speaking, any class that implements the `Comparable` interface and violates this condition should clearly indicate this fact.  The recommended language is 
> `"Note: this class has a natural ordering that is inconsistent with equals."`

but if you want to use this exit, please make it clear why we need to do so.

rpetry replied:
Thanks - I see that you have already fixed this.

alf239 commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/BestDesignElementCandidate.java:
```diff
@@ -0,0 +1,110 @@
+            return tStatDiff;
+        }
+
+        if (isNaN(getPValue()) || getPValue() > 1) // NA pVal for this experiment
```
I would prefer throwing an exception on P > 1: it simply doesn't make sense, sorting on the meaningless values isn't going to get us anywhere, and makes core cluttered with useless checks.

rpetry replied:
Is this comment still valid? You've made the change to this class but chose not to throw an exception for pval > 1?

alf239 replied:
I've replaced P > 1 with `NaN`; sure it's not the best (or a reasonable) way—I was eager to find out what else could be broken. 

I'd prefer not to have any candidates with `NaN`s, skipping incorrect values even before we create an instance. Will elaborate later today.

alf239 commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/BestDesignElementCandidate.java:
```diff
@@ -0,0 +1,110 @@
+        if (isNaN(getPValue()) || getPValue() > 1) // NA pVal for this experiment
+            return 1; // the other BestDesignElementCandidate comes first
+
+        if (isNaN(o.getPValue()) || o.getPValue() > 1) // NA pVal for the compared experiment
```
Note also the javadoc doesn't say anything about `NaN` or out-of-range P values.

rpetry replied:
Amended the javadoc.

alf239 commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/BestDesignElementCandidate.java:
```diff
@@ -0,0 +1,110 @@
+    private Float pValue;
+    private Float tStat;
+    private Integer deIndex;
+    private Integer uEFVIndex;
```
Unless you are absolutely sure you need to handle `null` values, avoid wrapper types. The only line where you use it,

``` java
BestDesignElementCandidate current = new BestDesignElementCandidate(pvals.get(i, j), tstat.get(i, j), i, j);
```

operates with proper primitives. Why do you need wrappers?

rpetry replied:
Changed wrappers to primitives

alf239 commented on atlas-web/src/main/java/ae3/service/experiment/AtlasExperimentAnalyticsViewService.java:
```diff
@@ -1,188 +1,155 @@
-        return findBestGenesForExperiment(expPart, geneIds, factors, factorValues, upDownCondition, offset, limit);
-    }
-
-    private BestDesignElementsResult findBestGenesForExperiment(
```
`atlas-web` is the last module where I would expect to see this method. What's so web-specific in it? What was wrong with `atlas-analytics`?

rpetry replied:
Well, I originally had this method in ExperimentWithData because to me it looks like a ncdf data access method. But you will see above that Olga sees this method as part of business logic (that I'd say belongs in atlas-web). Prior to this change (also in atlas-web) AtlasExperimentAnalyticsViewService contained code calling R functions.

alf239 committed:
Tests for compareTo are updated and partially fixed.

alf239 commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/BestDesignElementCandidate.java:
```diff
@@ -0,0 +1,114 @@
+     * 1. If absolute tStat ranks differ, return the higher tStat first
+     * 2. If absolute tStat ranks are the same:
+     * a. if pVals are different, return the lower pVal first
+     * b. if pVals are the same, return this BestDesignElementCandidate first
```
That means, a situation when `a < b` and `b < a` is perfectly possible. What was wrong with 0? (I've changed the code to use 0 when necessary, but I'm still wondering what was the rationale behind this decision.)

Oh, yes, the javadoc needs correction now.

rpetry replied:
Corrected javadoc

alf239 commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/ExperimentPart.java:
```diff
@@ -46,6 +50,18 @@ public ExperimentPart(ExperimentWithData ewd, ArrayDesign arrayDesign) {
+
+    public TwoDFloatArray getTStatistics() throws AtlasDataException, StatisticsNotFoundException {
+        return ewd.getTStatistics(arrayDesign);
+    }
```
As Olga pointed, the two methods above are _always_ called together. We all know the parallel arrays are bad, and you have even assigned a bug to me, asking to _encapsulate parallel arrays_—why do we add more parallel structures at the very same time?

rpetry replied:
Because I don't want to spend time re-inventing the solution you will implement across the whole of Atlas - unless you have a ready-made suggestion that will fit neatly into your solution.

alf239 replied:
fair enough

alf239 committed:
More checks, more fixes.

alf239 committed:
More checks, one test (NaN T statistics) is still failing

alf239 committed:
Even more checks - and more fails: x.compareTo(x) == 1 relation turns out to be non-reflexive

alf239 commented on atlas-web/src/main/java/ae3/service/experiment/AtlasExperimentAnalyticsViewService.java:
```diff
@@ -1,188 +1,155 @@
 
 /**
- * This class provides access to the statistical functions defined in R scripts.
+ * This class us used to populate the best genes table on the experiment page
```
a typo

rpetry replied:
Corrected.

rpetry committed:
Ticket #3215 - Code review comments

alf239 commented:
Olga has found an awfully stupid error in my tests — fixing, looking at the results :)

alf239 committed:
"reflexiveness" means for any x, `x.compareTo(x) == 0`, not  1 :)

alf239 commented:
Tests are still not passing, though:

```
a=BestDesignElementCandidate{pValue=NaN, tStat=NaN, deIndex=456339010, uEFVIndex=866036246}
b=BestDesignElementCandidate{pValue=0.0, tStat=Infinity, deIndex=-493540738, uEFVIndex=-1590161057}

java.lang.AssertionError: NaNs come last
    at org.junit.Assert.fail(Assert.java:91)
    at org.junit.Assert.assertTrue(Assert.java:43)
    at uk.ac.ebi.gxa.data.BestDesignElementCandidateTest.testTOrderingSemantics(BestDesignElementCandidateTest.java:99)
```

alf239 committed:
debug print is only necessary during debug

alf239 committed:
make it clear'n'easy

rpetry committed:
Ticket #3215 - Code review comments

alf239 commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/BestDesignElementCandidate.java:
```diff
@@ -0,0 +1,134 @@
+        }
+
+        if (isNaN(getPValue())) {
+            return isNaN(o.getPValue()) ? 0 : 1;
```
This line should not be reachable at the moment. Why do we keep it around?

rpetry replied:
Removed - compareTo() method looks so much simpler now

alf239 commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/BestDesignElementCandidate.java:
```diff
@@ -0,0 +1,134 @@
+        if (isNaN(getPValue())) {
+            return isNaN(o.getPValue()) ? 0 : 1;
+        } else if (isNaN(o.getPValue()))
+            return -1;
```
Same here.

rpetry replied:
Done

alf239 commented on atlas-data-storage/src/test/java/uk/ac/ebi/gxa/data/BestDesignElementCandidateGenerators.java:
```diff
@@ -0,0 +1,55 @@
+                                de.next(), uefv.next());
+                    } catch (UnexpectedException ue) {
+                        if (BestDesignElementCandidate.isPvalValid(pVal) && BestDesignElementCandidate.isTStatValid(tStat))
+                            throw LogUtil.createUnexpected("At least one of pVal:  " + pVal + "; tstat: " + tStat + " should be invalid");
```
We _explicitly_ ask to generate `NaN`s, so we can avoid this `catch` just by choosing "proper" value set. The check it self can be checked in fairly isolated manner, outside of this generator.

rpetry replied:
Removed

alf239 commented on atlas-data-storage/src/main/java/uk/ac/ebi/gxa/data/BestDesignElementCandidate.java:
```diff
@@ -0,0 +1,134 @@
+
+    public static boolean isTStatValid(float tStat) {
+        if (Float.isNaN(tStat))
+            return false;
```
Hint: 

![IDEA highligths it for you](https://img.skitch.com/20111212-km36f1q1heckrwiwk18re5acmj.jpg)

rpetry committed:
Ticket #3215 - More code review comments

rpetry committed:
Ticket #3215 - Fixed tests for invalid values.

Commit:
Moved the actual testing event to the test code

Generators are not supposed to be crashing, for most programmers it would mean a bug in the tests. Test should prepare the environment, and the failure must be clearly attributed to a given test.

Commit:
Integer -> int

Commit:
always comparing "this" with "other" to avoid confusion

rpetry committed:
Merge branch 'develop' into trac3215

Commit:
ANY_CONDITION condition is smart enough to know it fits everything—there's no need to check it explicitly

Commit:
The meaning of "allowed expression values" is better stored in UpDownExpression itself

After all, we have had a similar check there already

Commit:
More concise description of what constitutes a "valid P"

Commit:
fixed test: UpDownExpression is NA when p value is out of range [0, 1]

Commit:
More precise description of what constitutes an "invalid P"

Commit:
Narrow down the variable scope

Commit:
Dropped manual counting: we have `subList`

Commit:
Dropped now-obvious comment

rpetry committed:
Ticket #3215 - Allowed search not only by ef-efv, but also by ef only (search by gene id/factor is part of the experiment page functionality)

rpetry committed:
Merge branch 'trac3215' of github.com:gxa/gxa into trac3215

Commit:
Cleaned up `bestDesignElements`; dropped `KeyValuePair`

alf239 commented on atlas-model/src/test/java/uk/ac/ebi/microarray/atlas/model/UpDownExpressionTest.java:
```diff
@@ -59,27 +57,27 @@ public void testDownExpression() {
         assertTrue(isNonDe(0.05f, 0.0f));
         assertTrue(isNonDe(0.04f, 0.0f));
-        assertTrue(isNonDe(-0.04f, 0.0f));
+        assertTrue(isNA(-0.04f, 0.0f));
```
Don't we want to switch to QuickCheck? Too much effort for now?

alf239 commented on atlas-utils/src/main/java/uk/ac/ebi/gxa/utils/Best.java:
```diff
@@ -20,33 +20,20 @@
+/**
+ * @author alf
+ */
+public class Best<T extends Comparable<T>> {
```
It's not a change actually — Git's artefact

Commit:
Fixing the `ExperimentPartCriteria` initialization

Commit:
Oops. Rolling back—it was a mutator

Commit:
Rephrased with pairs

Commit:
Naming

Commit:
Proper use of `oneOf`

Commit:
Javadoc, ?:

Commit:
naming...

alf239 committed:
readability...

alf239 committed:
an awful bug

alf239 committed:
Predicates for UpDownExpression and gene ids; unused code eliminated where found

alf239 committed:
a NXOR b <=> a == b

alf239 committed:
We don't like NA any more

alf239 committed:
Unused logger

alf239 committed:
Notes on `compareTo` vs `equals`

alf239 committed:
inlined a method

alf239 committed:
made it more readable...

Commit:
DE candidate knows its t/p stats already

Commit:
proper immutability

Commit:
unused parts

Commit:
javadoc

Commit:
inlined gene id handling; made lines shorter

Commit:
"Candidate" -> "Statistics"; "Best" -> `Comparable`

Commit:
Extracted the utility method

Commit:
o -> x

Commit:
o -> x

Commit:
Getting rid of the Best
